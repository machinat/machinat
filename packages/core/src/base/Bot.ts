import type { SociablyBot, SociablyThread, SociablyNode } from '../types.js';
import type { DispatchResponse } from '../engine/types.js';
import { serviceInterface, serviceProviderClass } from '../service/index.js';

type BaseBotI = SociablyBot<SociablyThread, unknown, unknown>;

const BotPlatformMapI = serviceInterface<BaseBotI>({
  name: 'BotPlatformMap',
  polymorphic: true,
});

/**
 * @category Base
 */
export class BaseBot implements BaseBotI {
  static PlatformMap = BotPlatformMapI;

  private _platformMapping: Map<string, BaseBotI>;

  constructor(platformMapping: Map<string, BaseBotI>) {
    this._platformMapping = platformMapping;
  }

  async render(
    thread: SociablyThread,
    node: SociablyNode
  ): Promise<null | DispatchResponse<unknown, unknown>> {
    const bot = this._platformMapping.get(thread.platform);
    if (!bot) {
      throw new TypeError(
        `thread of platform '${thread.platform}' is not supported`
      );
    }

    return bot.render(thread, node);
  }

  async start(): Promise<void> {} // eslint-disable-line class-methods-use-this
  async stop(): Promise<void> {} // eslint-disable-line class-methods-use-this
}

const BotP = serviceProviderClass({
  lifetime: 'transient',
  deps: [BotPlatformMapI],
})(BaseBot);

type BotP = BaseBot;

export default BotP;
