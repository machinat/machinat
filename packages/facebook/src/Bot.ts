import { AgentSettingsAccessor } from '@sociably/core';
import Engine, { DispatchError } from '@sociably/core/engine';
import Queue from '@sociably/core/queue';
import Renderer from '@sociably/core/renderer';
import ModuleUtilitiesI from '@sociably/core/base/ModuleUtilities';
import { serviceProviderClass } from '@sociably/core/service';
import type {
  SociablyNode,
  SociablyBot,
  InitScopeFn,
  DispatchWrapper,
} from '@sociably/core';
import {
  MetaApiWorker,
  MetaApiJob,
  MetaApiResult,
  MetaApiDispatchResponse,
  MetaApiResponseBody,
} from '@sociably/meta-api';
import generalComponentDelegator from './components/general';
import { FACEBOOK, PATH_FEED, PATH_PHOTOS } from './constant';
import {
  ConfigsI,
  PlatformUtilitiesI,
  PageSettingsAccessorI,
} from './interface';
import FacebookPage from './Page';
import FacebookChat from './Chat';
import InteractTarget from './InteractTarget';
import {
  createChatJobs,
  createChatAttachmentJobs,
  createPostJobs,
  createInteractJobs,
} from './job';
import type {
  FacebookThread,
  FacebookComponent,
  FacebookSegmentValue,
  FacebookDispatchFrame,
  MessagingOptions,
  FacebookPageSettings,
} from './types';

type FacebookDispatchTarget = FacebookThread | FacebookPage;

type FacebookBotOptions = {
  appSecret?: string;
  graphApiVersion?: string;
  apiBatchRequestInterval?: number;
  pageSettingsAccessor: AgentSettingsAccessor<
    FacebookPage,
    FacebookPageSettings
  >;
  initScope?: InitScopeFn;
  dispatchWrapper?: DispatchWrapper<
    MetaApiJob,
    FacebookDispatchFrame,
    MetaApiResult
  >;
};

type UploadAttachmentResult = {
  attachmentId: string;
};

type PagePhotoResult = {
  photoId: string;
};

type PagePostResult = {
  postId: string;
  photos: PagePhotoResult[];
};

type CommentResult = {
  commentId: string;
  photo: null | PagePhotoResult;
};

type ApiCallOptions = {
  /** The page to make the API call */
  page: string | FacebookPage;
  /** HTTP method */
  method?: string;
  /** API request URL relative to https://graph.facebook.com/{version}/ */
  url: string;
  /** API request parameters */
  params?: Record<string, unknown>;
};

/**
 * FacebookBot render messages and make API call to Facebook platform.
 * @category Provider
 */
export class FacebookBot
  implements SociablyBot<FacebookThread, MetaApiJob, MetaApiResult>
{
  worker: MetaApiWorker;
  engine: Engine<
    FacebookDispatchTarget,
    FacebookSegmentValue,
    FacebookComponent<unknown>,
    MetaApiJob,
    MetaApiResult
  >;

  platform = FACEBOOK;

  constructor({
    appSecret,
    graphApiVersion = 'v11.0',
    apiBatchRequestInterval = 500,
    pageSettingsAccessor,
    initScope,
    dispatchWrapper,
  }: FacebookBotOptions) {
    const renderer = new Renderer<
      FacebookSegmentValue,
      FacebookComponent<unknown>
    >(FACEBOOK, generalComponentDelegator);

    const queue = new Queue<MetaApiJob, MetaApiResult>();
    const worker = new MetaApiWorker(
      pageSettingsAccessor,
      appSecret,
      graphApiVersion,
      apiBatchRequestInterval
    );

    this.engine = new Engine(
      FACEBOOK,
      renderer,
      queue,
      worker,
      initScope,
      dispatchWrapper
    );
  }

  async start(): Promise<void> {
    this.engine.start();
  }

  async stop(): Promise<void> {
    this.engine.stop();
  }

  async render(
    target: FacebookDispatchTarget,
    node: SociablyNode
  ): Promise<null | MetaApiDispatchResponse> {
    if (target instanceof FacebookChat) {
      return this.engine.render(target, node, createChatJobs());
    }
    if (target instanceof FacebookPage) {
      return this.engine.render(target, node, createPostJobs);
    }
    if (target instanceof InteractTarget) {
      return this.engine.render(target, node, createInteractJobs);
    }
    throw new TypeError('invalid rendering target');
  }

  /** Send messages or actions on a chat */
  async message(
    chat: FacebookChat,
    messages: SociablyNode,
    options?: MessagingOptions
  ): Promise<null | MetaApiDispatchResponse> {
    return this.engine.render(chat, messages, createChatJobs(options));
  }

  /** Upload a media chat attachment for later use */
  async uploadChatAttachment(
    /** The {@link FacebookPage} that owns the attachment */
    pageInput: string | FacebookPage,
    /** An {@link Image}, {@link Audio}, {@link Video} or {@link File} element to be uploaded */ node: SociablyNode
  ): Promise<null | UploadAttachmentResult> {
    const page =
      typeof pageInput === 'string' ? new FacebookPage(pageInput) : pageInput;
    const response = await this.engine.render(
      page,
      node,
      createChatAttachmentJobs
    );
    const result = response?.results[0].body;
    return result ? { attachmentId: result.attachment_id } : null;
  }

  /** Create a post or a photo on the page feed */
  async post(
    /** The {@link FacebookPage} to post */
    pageInput: string | FacebookPage,
    /** Text, a {@link PagePost} or a {@link PagePhoto} to post */
    node: SociablyNode
  ): Promise<null | PagePostResult> {
    const page =
      typeof pageInput === 'string' ? new FacebookPage(pageInput) : pageInput;
    const response = await this.engine.render(page, node, createPostJobs);
    if (!response) {
      return null;
    }

    const photos: PagePhotoResult[] = [];
    let postId: undefined | string;

    const { jobs, results } = response;
    for (const [i, { request }] of jobs.entries()) {
      const resultBody = results[i].body;

      if (request.url === PATH_FEED) {
        postId = resultBody.id;
      } else if (request.url === PATH_PHOTOS) {
        if (resultBody.post_id) {
          postId = resultBody.post_id;
        }
        photos.push({ photoId: resultBody.id });
      }
    }
    if (!postId) {
      return null;
    }
    return { postId, photos };
  }

  /** Comment or make reactions to a Facebook post or comment */
  async interact(
    /** The target thread to interact with */
    target: InteractTarget,
    /** {@link Comment} or {@link Reaction} */
    node: SociablyNode
  ): Promise<null | CommentResult[]> {
    const response = await this.engine.render(target, node, createInteractJobs);
    if (!response) {
      return null;
    }

    const results: CommentResult[] = [];
    let photo: null | PagePhotoResult = null;

    for (const [i, { request }] of response.jobs.entries()) {
      if (request.url === PATH_PHOTOS) {
        photo = { photoId: response.results[i].body.id };
      } else if (request.url === PATH_PHOTOS) {
        results.push({
          commentId: response.results[i].body.id,
          photo,
        });
        photo = null;
      }
    }
    return results;
  }

  async requestApi<ResBody extends MetaApiResponseBody>({
    page: pageInput,
    method = 'GET',
    url,
    params,
  }: ApiCallOptions): Promise<ResBody> {
    const page =
      typeof pageInput === 'string' ? new FacebookPage(pageInput) : pageInput;
    try {
      const { results } = await this.engine.dispatchJobs(page, [
        {
          channel: page,
          request: { method, url, params },
        },
      ]);

      return results[0].body as ResBody;
    } catch (err) {
      if (err instanceof DispatchError) {
        throw err.errors[0];
      }
      throw err;
    }
  }
}

const BotP = serviceProviderClass({
  lifetime: 'singleton',
  deps: [
    ConfigsI,
    PageSettingsAccessorI,
    { require: ModuleUtilitiesI, optional: true },
    { require: PlatformUtilitiesI, optional: true },
  ],
  factory: (
    { appSecret, apiBatchRequestInterval },
    pageSettingsAccessor,
    moduleUitils,
    platformUtils
  ) =>
    new FacebookBot({
      pageSettingsAccessor,
      appSecret,
      apiBatchRequestInterval,
      initScope: moduleUitils?.initScope,
      dispatchWrapper: platformUtils?.dispatchWrapper,
    }),
})(FacebookBot);

type BotP = FacebookBot;
export default BotP;
