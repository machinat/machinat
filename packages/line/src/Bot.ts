import invariant from 'invariant';
import type {
  SociablyNode,
  SociablyBot,
  InitScopeFn,
  DispatchWrapper,
} from '@sociably/core';
import Renderer from '@sociably/core/renderer';
import Queue from '@sociably/core/queue';
import Engine, { DispatchError } from '@sociably/core/engine';
import ModuleUtilitiesI from '@sociably/core/base/ModuleUtilities';
import { makeClassProvider } from '@sociably/core/service';

import { createChatJobs, createMulticastJobs } from './job';
import generalElementDelegate from './components/general';
import LineWorker from './Worker';
import LineChat from './Chat';
import { ConfigsI, PlatformUtilitiesI } from './interface';
import { LINE } from './constant';
import type {
  LineSource,
  LineSegmentValue,
  LineComponent,
  LineJob,
  LineResult,
  LineDispatchFrame,
  LineDispatchResponse,
  MessagingApiResult,
} from './types';

type LineBotOptions = {
  providerId: string;
  channelId: string;
  accessToken: string;
  maxRequestConnections?: number;
  initScope?: InitScopeFn;
  dispatchWrapper?: DispatchWrapper<LineJob, LineDispatchFrame, LineResult>;
};

/**
 * @category Provider
 */
export class LineBot implements SociablyBot<LineChat, LineJob, LineResult> {
  providerId: string;
  channelId: string;
  maxRequestConnections: number;
  engine: Engine<
    LineChat,
    LineSegmentValue,
    LineComponent<unknown>,
    LineJob,
    LineResult
  >;

  platform = LINE;

  constructor({
    providerId,
    channelId,
    accessToken,
    maxRequestConnections = 100,
    initScope,
    dispatchWrapper,
  }: LineBotOptions) {
    invariant(accessToken, 'configs.accessToken should not be empty');
    invariant(providerId, 'configs.providerId should not be empty');
    invariant(channelId, 'configs.channelId should not be empty');

    this.providerId = providerId;
    this.channelId = channelId;
    this.maxRequestConnections = maxRequestConnections;

    const queue = new Queue<LineJob, LineResult>();
    const worker = new LineWorker(accessToken, maxRequestConnections);
    const renderer = new Renderer<LineSegmentValue, LineComponent<unknown>>(
      LINE,
      generalElementDelegate
    );

    this.engine = new Engine(
      LINE,
      renderer,
      queue,
      worker,
      initScope,
      dispatchWrapper
    );
  }

  async start(): Promise<void> {
    this.engine.start();
  }

  async stop(): Promise<void> {
    this.engine.stop();
  }

  render(
    source: string | LineSource | LineChat,
    message: SociablyNode,
    options?: { replyToken?: string }
  ): Promise<null | LineDispatchResponse> {
    const thread =
      source instanceof LineChat
        ? source
        : typeof source === 'string'
        ? new LineChat(this.channelId, 'user', source)
        : LineChat.fromMessagingSource(this.channelId, source);

    return this.engine.render(
      thread,
      message,
      createChatJobs(options && options.replyToken)
    );
  }

  renderMulticast(
    targets: string[],
    message: SociablyNode
  ): Promise<null | LineDispatchResponse> {
    return this.engine.render(null, message, createMulticastJobs(targets));
  }

  async makeApiCall<ResBody extends MessagingApiResult>(
    method: 'GET' | 'POST' | 'PUT' | 'DELETE',
    path: string,
    body?: unknown
  ): Promise<ResBody> {
    try {
      const response = await this.engine.dispatchJobs(null, [
        { method, path, body, executionKey: undefined },
      ]);

      return response.results[0].body as ResBody;
    } catch (err) {
      if (err instanceof DispatchError) {
        throw err.errors[0];
      }
      throw err;
    }
  }
}

const BotP = makeClassProvider({
  lifetime: 'singleton',
  deps: [
    ConfigsI,
    { require: ModuleUtilitiesI, optional: true },
    { require: PlatformUtilitiesI, optional: true },
  ],
  factory: (
    { providerId, channelId, accessToken, maxRequestConnections },
    moduleUtils,
    platformUtils
  ) =>
    new LineBot({
      providerId,
      channelId,
      accessToken,
      maxRequestConnections,
      initScope: moduleUtils?.initScope,
      dispatchWrapper: platformUtils?.dispatchWrapper,
    }),
})(LineBot);

type BotP = LineBot;

export default BotP;
