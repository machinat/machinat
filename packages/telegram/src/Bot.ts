import invariant from 'invariant';
import fetch from 'node-fetch';
import type {
  MachinatNode,
  MachinatBot,
  InitScopeFn,
  DispatchWrapper,
} from '@machinat/core';
import Renderer from '@machinat/core/renderer';
import Queue from '@machinat/core/queue';
import Engine, { DispatchError } from '@machinat/core/engine';
import ModuleUtilitiesI from '@machinat/core/base/ModuleUtilities';
import { makeClassProvider } from '@machinat/core/service';

import { createChatJob, createDirectInstanceJobs } from './job';
import generalElementDelegate from './components/general';
import TelegramWorker from './Worker';
import TelegramChat from './Chat';
import TelegramChatTarget from './ChatTarget';
import { ConfigsI, PlatformUtilitiesI } from './interface';
import { TELEGRAM } from './constant';
import TelegramApiError from './Error';
import type {
  TelegramSegmentValue,
  TelegramJob,
  TelegramResult,
  BotApiResult,
  TelegramDispatchFrame,
  TelegramDispatchResponse,
  UploadingFile,
  TelegramComponent,
} from './types';

type TelegramBotOptions = {
  token: string;
  maxConnections?: number;
  initScope?: InitScopeFn;
  dispatchWrapper?: DispatchWrapper<
    TelegramJob,
    TelegramDispatchFrame,
    TelegramResult
  >;
};

/**
 * @category Provider
 */
export class TelegramBot
  implements
    MachinatBot<TelegramChat | TelegramChatTarget, TelegramJob, TelegramResult>
{
  token: string;
  id: number;
  engine: Engine<
    TelegramChat | TelegramChatTarget,
    TelegramSegmentValue,
    TelegramComponent<unknown>,
    TelegramJob,
    TelegramResult
  >;

  constructor({
    token,
    maxConnections = 100,
    initScope,
    dispatchWrapper,
  }: TelegramBotOptions) {
    invariant(token, 'options.token should not be empty');

    this.token = token;
    this.id = Number(token.split(':', 1)[0]);

    const queue = new Queue<TelegramJob, TelegramResult>();
    const worker = new TelegramWorker(token, maxConnections);
    const renderer = new Renderer<
      TelegramSegmentValue,
      TelegramComponent<unknown>
    >(TELEGRAM, generalElementDelegate);

    this.engine = new Engine(
      TELEGRAM,
      renderer,
      queue,
      worker,
      initScope,
      dispatchWrapper
    );
  }

  async start(): Promise<void> {
    this.engine.start();
  }

  async stop(): Promise<void> {
    this.engine.stop();
  }

  render(
    target: number | string | TelegramChat | TelegramChatTarget,
    message: MachinatNode
  ): Promise<null | TelegramDispatchResponse> {
    const channel =
      target instanceof TelegramChat || target instanceof TelegramChatTarget
        ? target
        : new TelegramChatTarget(this.id, target);

    return this.engine.render(channel, message, createChatJob);
  }

  /**
   * Render specific instance directly without target chat, only allow
   * {@link AnswerCallbackQuery}, {@link AnswerInlineQuery},
   * {@link AnswerShippingQuery}, {@link AnswerPreCheckoutQuery} and editing
   * inline mode messages (using {@link EditText}, {@link EditCaption} and
   * {@link EditMedia} with inlineMessageId prop).
   */
  renderInstance(
    message: MachinatNode
  ): Promise<null | TelegramDispatchResponse> {
    return this.engine.render(null, message, createDirectInstanceJobs);
  }

  async fetchFile(fileId: string): Promise<null | {
    content: NodeJS.ReadableStream;
    contentType?: string;
    contentLength?: number;
  }> {
    const { file_path: filePath } = await this.makeApiCall('getFile', {
      file_id: fileId,
    });

    if (!filePath) {
      return null;
    }

    const fetchResponse = await fetch(
      `https://api.telegram.org/file/bot${this.token}/${filePath}`
    );

    if (!fetchResponse.ok) {
      throw new TelegramApiError({
        ok: false,
        description: fetchResponse.statusText,
        error_code: fetchResponse.status,
      });
    }

    const contentLength = fetchResponse.headers.get('content-length');
    return {
      content: fetchResponse.body,
      contentType: fetchResponse.headers.get('content-type') || undefined,
      contentLength: contentLength ? Number(contentLength) : undefined,
    };
  }

  async makeApiCall<Result extends BotApiResult>(
    method: string,
    parameters: Record<string, unknown> = {},
    uploadingFiles?: UploadingFile[]
  ): Promise<Result> {
    try {
      const response = await this.engine.dispatchJobs(null, [
        {
          method,
          parameters,
          executionKey: undefined,
          uploadingFiles: uploadingFiles || null,
        },
      ]);

      return response.results[0].result as Result;
    } catch (err) {
      if (err instanceof DispatchError) {
        throw err.errors[0];
      }
      throw err;
    }
  }
}

const BotP = makeClassProvider({
  lifetime: 'singleton',
  deps: [
    ConfigsI,
    { require: ModuleUtilitiesI, optional: true },
    { require: PlatformUtilitiesI, optional: true },
  ],
  factory: ({ botToken, maxConnections }, moduleUtils, platformUtils) => {
    invariant(botToken, 'configs.botToken should not be empty');

    return new TelegramBot({
      token: botToken,
      maxConnections,
      initScope: moduleUtils?.initScope,
      dispatchWrapper: platformUtils?.dispatchWrapper,
    });
  },
})(TelegramBot);

type BotP = TelegramBot;
export default BotP;
