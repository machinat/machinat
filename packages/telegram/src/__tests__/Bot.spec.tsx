/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { Readable } from 'stream';
import moxy, { Moxy } from '@moxyjs/moxy';
import nock from 'nock';
import Sociably from '@sociably/core';
import Queue from '@sociably/core/queue';
import _Engine from '@sociably/core/engine';
import _Renderer from '@sociably/core/renderer';
import _Worker from '../Worker';
import TelegramChat from '../Chat';
import TelegramUser from '../User';
import TelegramApiError from '../Error';
import { TelegramBot } from '../Bot';
import {
  Photo,
  Expression,
  ForceReply,
  EditText,
  EditMedia,
} from '../components';

const Engine = _Engine as Moxy<typeof _Engine>;
const Renderer = _Renderer as Moxy<typeof _Renderer>;
const Worker = _Worker as Moxy<typeof _Worker>;

nock.disableNetConnect();

jest.mock('@sociably/core/engine', () =>
  jest
    .requireActual('@moxyjs/moxy')
    .default(jest.requireActual('@sociably/core/engine'))
);

jest.mock('@sociably/core/renderer', () =>
  jest
    .requireActual('@moxyjs/moxy')
    .default(jest.requireActual('@sociably/core/renderer'))
);

jest.mock('../Worker', () =>
  jest.requireActual('@moxyjs/moxy').default(jest.requireActual('../Worker'), {
    mockNewInstance: false,
  })
);

const initScope = moxy(() => moxy());
const dispatchWrapper = moxy((x) => x);

const agentSettings = {
  botToken: '12345:_BOT_TOKEN_',
  botName: 'MyBot',
  secretToken: '_SECRET_TOKEN_',
};

const agentSettingsAccessor = moxy({
  getAgentSettings: async () => agentSettings,
  getAgentSettingsBatch: async () => [],
});

const botUser = new TelegramUser(12345, true);

const message = (
  <Expression replyMarkup={<ForceReply />}>
    Hello <b>World!</b>
    <Photo url="https://sociably.io/greeting.png" />
  </Expression>
);

const telegramApi = nock('https://api.telegram.org');
const bodySpy = moxy(() => true);

beforeEach(() => {
  nock.cleanAll();
  bodySpy.mock.clear();
  Engine.mock.clear();
  Renderer.mock.clear();
  Worker.mock.clear();
});

describe('.constructor(options)', () => {
  it('assemble core modules', () => {
    const bot = new TelegramBot({
      initScope,
      dispatchWrapper,
      agentSettingsAccessor,
      maxRequestConnections: 999,
    });

    expect(bot.engine).toBeInstanceOf(Engine);

    expect(Renderer).toHaveBeenCalledTimes(1);
    expect(Renderer).toHaveBeenCalledWith('telegram', expect.any(Function));

    expect(Engine).toHaveBeenCalledTimes(1);
    expect(Engine).toHaveBeenCalledWith(
      'telegram',
      expect.any(Renderer),
      expect.any(Queue),
      expect.any(Worker),
      initScope,
      dispatchWrapper
    );

    expect(Worker).toHaveBeenCalledTimes(1);
    expect(Worker).toHaveBeenCalledWith(agentSettingsAccessor, 999);
  });

  test('default maxRequestConnections', () => {
    expect(
      new TelegramBot({
        initScope,
        dispatchWrapper,
        agentSettingsAccessor,
      })
    );

    expect(Worker).toHaveBeenCalledTimes(1);
    expect(Worker.mock.calls[0].args).toMatchInlineSnapshot(`
      Array [
        Object {
          "getAgentSettings": [MockFunction moxy(getAgentSettings)],
          "getAgentSettingsBatch": [MockFunction moxy(getAgentSettingsBatch)],
        },
        100,
      ]
    `);
  });
});

test('.start() and .stop() start/stop engine', () => {
  const bot = new TelegramBot({
    initScope,
    dispatchWrapper,
    agentSettingsAccessor,
  });

  type MockEngine = Moxy<TelegramBot['engine']>;

  bot.start();
  expect((bot.engine as MockEngine).start).toHaveBeenCalledTimes(1);

  bot.stop();
  expect((bot.engine as MockEngine).stop).toHaveBeenCalledTimes(1);
});

describe('.render(thread, message, options)', () => {
  const bot = new TelegramBot({ agentSettingsAccessor });

  beforeAll(() => {
    bot.start();
  });

  afterAll(() => {
    bot.stop();
  });

  it('resolves null if message is empty', async () => {
    const empties = [undefined, null, [], <></>];
    for (const empty of empties) {
      // eslint-disable-next-line no-await-in-loop
      await expect(
        bot.render(new TelegramChat(12345, 67890), empty)
      ).resolves.toBe(null);
    }
  });

  it('send rendered messages to telegram api', async () => {
    const apiCall1 = telegramApi
      .post('/bot12345:_BOT_TOKEN_/sendMessage', bodySpy)
      .reply(200, { ok: true, result: { id: 1 } });
    const apiCall2 = telegramApi
      .post('/bot12345:_BOT_TOKEN_/sendPhoto', bodySpy)
      .reply(200, { ok: true, result: { id: 2 } });

    const response = await bot.render(new TelegramChat(12345, 67890), message);
    expect(response).toMatchSnapshot();

    expect(response!.results).toEqual([
      { ok: true, result: { id: 1 } },
      { ok: true, result: { id: 2 } },
    ]);

    expect(bodySpy).toHaveBeenCalledTimes(2);
    expect(bodySpy.mock.calls[0].args[0]).toMatchInlineSnapshot(`
      Object {
        "chat_id": 67890,
        "parse_mode": "HTML",
        "text": "Hello <b>World!</b>",
      }
    `);
    expect(bodySpy.mock.calls[1].args[0]).toMatchInlineSnapshot(`
      Object {
        "chat_id": 67890,
        "parse_mode": "HTML",
        "photo": "https://sociably.io/greeting.png",
        "reply_markup": Object {
          "force_reply": true,
        },
      }
    `);

    expect(apiCall1.isDone()).toBe(true);
    expect(apiCall2.isDone()).toBe(true);
  });

  test('render global actions without a chat target', async () => {
    const apiCall1 = telegramApi
      .post('/bot12345:_BOT_TOKEN_/editMessageText', bodySpy)
      .reply(200, { ok: true, result: { id: '1' } });
    const apiCall2 = telegramApi
      .post('/bot12345:_BOT_TOKEN_/editMessageMedia', bodySpy)
      .reply(200, { ok: true, result: { id: '2' } });

    const response = await bot.render(
      new TelegramUser(12345, true),
      <>
        <EditText inlineMessageId="1">
          foo <b>bar</b>
        </EditText>
        <EditMedia inlineMessageId="2">
          <Photo url="https://sociably.io/trollface.png" />
        </EditMedia>
      </>
    );
    expect(response!).toMatchSnapshot();
    expect(response!.results).toEqual([
      { ok: true, result: { id: '1' } },
      { ok: true, result: { id: '2' } },
    ]);

    expect(bodySpy).toHaveBeenCalledTimes(2);
    expect(bodySpy.mock.calls[0].args[0]).toMatchInlineSnapshot(`
      Object {
        "inline_message_id": "1",
        "parse_mode": "HTML",
        "text": "foo <b>bar</b>",
      }
    `);
    expect(bodySpy.mock.calls[1].args[0]).toMatchInlineSnapshot(`
      Object {
        "inline_message_id": "2",
        "media": Object {
          "media": "https://sociably.io/trollface.png",
          "parse_mode": "HTML",
          "type": "photo",
        },
      }
    `);

    expect(apiCall1.isDone()).toBe(true);
    expect(apiCall2.isDone()).toBe(true);
  });

  test('render empty message when no chat target', async () => {
    const empties = [undefined, null, [], <></>];
    for (const empty of empties) {
      // eslint-disable-next-line no-await-in-loop
      await expect(
        bot.render(new TelegramUser(12345, true), empty)
      ).resolves.toBe(null);
    }
  });
});

describe('.requestApi()', () => {
  test('call telegram bot api', async () => {
    const bot = new TelegramBot({
      initScope,
      dispatchWrapper,
      agentSettingsAccessor,
    });
    bot.start();

    const fooCall = telegramApi
      .post('/bot12345:_BOT_TOKEN_/foo', bodySpy)
      .reply(200, {
        ok: true,
        result: { foo: 'bar' },
      });

    await expect(
      bot.requestApi({ agent: botUser, method: 'foo', params: { bar: 'baz' } })
    ).resolves.toEqual({
      foo: 'bar',
    });

    expect(bodySpy).toHaveBeenCalledTimes(1);
    expect(bodySpy).toHaveBeenCalledWith({ bar: 'baz' });

    expect(fooCall.isDone()).toBe(true);
  });

  it('throw TelegramAPIError when fail', async () => {
    const bot = new TelegramBot({
      initScope,
      dispatchWrapper,
      agentSettingsAccessor,
    });
    bot.start();

    const failBody = {
      ok: false,
      error: 'bad',
      error_description: 'really bad',
    };

    const fooCall = telegramApi
      .post('/bot12345:_BOT_TOKEN_/foo', bodySpy)
      .reply(200, failBody);

    try {
      await bot.requestApi({
        agent: botUser,
        method: 'foo',
        params: { bar: 'baz' },
      });
      expect('should not get here').toBeFalsy();
    } catch (error) {
      expect(error).toBeInstanceOf(TelegramApiError);
      expect(error.body).toEqual(failBody);
    }

    expect(bodySpy).toHaveBeenCalledTimes(1);
    expect(bodySpy).toHaveBeenCalledWith({ bar: 'baz' });

    expect(fooCall.isDone()).toBe(true);
  });
});

test('.fetchFile()', async () => {
  const bot = new TelegramBot({
    initScope,
    dispatchWrapper,
    agentSettingsAccessor,
  });
  bot.start();

  const fileId = '_FILE_ID_';

  const getFileCall = telegramApi
    .post('/bot12345:_BOT_TOKEN_/getFile', bodySpy)
    .reply(200, {
      ok: true,
      result: {
        file_id: fileId,
        file_unique_id: '_FILE_UNIQUE_ID_',
        file_path: '_FILE_PATH_',
      },
    });

  const downloadFileCall = telegramApi
    .get('/file/bot12345:_BOT_TOKEN_/_FILE_PATH_', bodySpy)
    .reply(200, '__BINARY_DATA__', {
      'Content-Type': 'image/png',
      'Content-Length': '777',
    });

  const response = await bot.fetchFile(botUser, fileId);

  expect(bodySpy).toHaveBeenCalledTimes(2);
  expect(bodySpy).toHaveBeenNthCalledWith(1, { file_id: fileId });

  expect(response!.content).toBeInstanceOf(Readable);
  expect(response!.contentType).toBe('image/png');
  expect(response!.contentLength).toBe(777);

  await new Promise(setImmediate);
  expect(response!.content.read().toString()).toBe('__BINARY_DATA__');

  expect(getFileCall.isDone()).toBe(true);
  expect(downloadFileCall.isDone()).toBe(true);
});
