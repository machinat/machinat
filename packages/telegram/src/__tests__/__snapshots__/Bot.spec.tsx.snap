// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#render(channel, message, options) send rendered messages to telegram api 1`] = `
Object {
  "jobs": Array [
    Object {
      "executionKey": "telegram.12345.123456",
      "method": "sendMessage",
      "parameters": Object {
        "chat_id": 123456,
        "disable_notification": undefined,
        "parse_mode": "HTML",
        "text": "Hello <b>World!</b>",
      },
      "uploadingFiles": null,
    },
    Object {
      "executionKey": "telegram.12345.123456",
      "method": "sendPhoto",
      "parameters": Object {
        "caption": undefined,
        "chat_id": 123456,
        "disable_notification": undefined,
        "parse_mode": "HTML",
        "photo": "https://machinat.com/greeting.png",
        "reply_markup": Object {
          "force_reply": true,
          "selective": undefined,
        },
        "reply_to_message_id": undefined,
      },
      "uploadingFiles": null,
    },
  ],
  "results": Array [
    Object {
      "ok": true,
      "result": Object {
        "id": 1,
      },
    },
    Object {
      "ok": true,
      "result": Object {
        "id": 2,
      },
    },
  ],
  "tasks": Array [
    Object {
      "payload": Array [
        Object {
          "executionKey": "telegram.12345.123456",
          "method": "sendMessage",
          "parameters": Object {
            "chat_id": 123456,
            "disable_notification": undefined,
            "parse_mode": "HTML",
            "text": "Hello <b>World!</b>",
          },
          "uploadingFiles": null,
        },
        Object {
          "executionKey": "telegram.12345.123456",
          "method": "sendPhoto",
          "parameters": Object {
            "caption": undefined,
            "chat_id": 123456,
            "disable_notification": undefined,
            "parse_mode": "HTML",
            "photo": "https://machinat.com/greeting.png",
            "reply_markup": Object {
              "force_reply": true,
              "selective": undefined,
            },
            "reply_to_message_id": undefined,
          },
          "uploadingFiles": null,
        },
      ],
      "type": "dispatch",
    },
  ],
}
`;

exports[`#renderInstance(message) call edit api 1`] = `
Object {
  "jobs": Array [
    Object {
      "executionKey": undefined,
      "method": "editMessageText",
      "parameters": Object {
        "disable_web_page_preview": undefined,
        "inline_message_id": "1",
        "message_id": undefined,
        "parse_mode": "HTML",
        "reply_markup": undefined,
        "text": "foo <b>bar</b>",
      },
      "uploadingFiles": null,
    },
    Object {
      "executionKey": undefined,
      "method": "editMessageMedia",
      "parameters": Object {
        "inline_message_id": "2",
        "media": Object {
          "caption": undefined,
          "media": "https://machinat.com/trollface.png",
          "parse_mode": "HTML",
          "type": "photo",
        },
        "message_id": undefined,
        "reply_markup": undefined,
      },
      "uploadingFiles": null,
    },
  ],
  "results": Array [
    Object {
      "ok": true,
      "result": Object {
        "id": "1",
      },
    },
    Object {
      "ok": true,
      "result": Object {
        "id": "2",
      },
    },
  ],
  "tasks": Array [
    Object {
      "payload": Array [
        Object {
          "executionKey": undefined,
          "method": "editMessageText",
          "parameters": Object {
            "disable_web_page_preview": undefined,
            "inline_message_id": "1",
            "message_id": undefined,
            "parse_mode": "HTML",
            "reply_markup": undefined,
            "text": "foo <b>bar</b>",
          },
          "uploadingFiles": null,
        },
        Object {
          "executionKey": undefined,
          "method": "editMessageMedia",
          "parameters": Object {
            "inline_message_id": "2",
            "media": Object {
              "caption": undefined,
              "media": "https://machinat.com/trollface.png",
              "parse_mode": "HTML",
              "type": "photo",
            },
            "message_id": undefined,
            "reply_markup": undefined,
          },
          "uploadingFiles": null,
        },
      ],
      "type": "dispatch",
    },
  ],
}
`;
